xla/service/gpu/transforms/windowed_einsum_handler.cc

```diff
@@ -492,6 +511,125 @@ absl::Status PostProcessUnrolledLoop(HloInstruction* loop, int64_t stream_id) {
       TF_RETURN_IF_ERROR(UpdateDotAndConsumerConfig(inst, stream_id));
       ++stream_id;
     }
+    // If dot's result is accumulated, this means we found a loop with
+    // contracting dim sharded.
+    HloInstruction* partial_dot;
+    if (Match(inst, m::AddAnyOrder(m::Op(),
+                                   m::Dot(&partial_dot, m::Op(), m::Op())))) {
+      partial_accumulations.push_back(partial_dot);
+    }
+  }
+
+  // Transform partial accumulations into a reduction on a contiguous buffer.
+  // Partial accumulations will impact the overlap between dots because the
+  // dot+add pattern will be fused into a single gemm later in gemm rewriter
+  // which adds data dependencies between gemms. Instead we write all
+  // intermediate results into a larger buffer and perform a one-shot reduction.
+  // The high-level transformation is:
+  // shape(x,y) cp   shape(x,y) dot0
+  //          \         /
+  //           \       /
+  //     shape(x,y) add0    shape(x,y) dot1
+  //             \                /
+  //              \              /
+  //             shape(x,y) add1
+  //                     |
+  //        shape(x,y) loop output
+  //
+  // transformed into:
+  // shape(x,y) cp         shape(x,y) dot0    shape(x,y) dot1
+  //    \                     /                   /
+  //     \                   /                   /
+  //     shape(n,x,y) concatenate on first axis, n is the number of partitions
+  //                        |
+  //             shape(n,x,y) loop output
+  //                        |
+  //             shape(x,y) reduction on first axis
+  //
+  // The final reduction is pulled outside of the loop to overlap with other
+  // collectives.
+  if (partial_accumulations.size() > 0 &&
+      while_body->name().find(
+          WindowedEinsumHandler::kWindowedEinsumAgLoopName) !=
+          std::string::npos) {
+    std::vector<HloInstruction*> partials_to_concat;
+
+    // We reshape it to a N+1 dimensioned tensor with left-most dim being 1.
+    Shape shape = partial_accumulations[0]->shape();
+    shape = ShapeUtil::PrependMajorDimension(1, shape);
+
+    for (auto& inst : partial_accumulations) {
+      CHECK(inst->user_count() == 1);
```

Comment:
I would not assume the user count is exactly 1. I think this is currently the case for SPMD partitioner, but we should not assume SPMD partitioner will always use a dot at most once. In general, we shouldn't assume too much about the precise output of SPMD partitioner since the output can change in the future.

Reply:
Yea agreed, will remove the check